File: client/audit_split_view.py
"""In-client AUDIT.md split view — device-only visit logs + pane refresh policy.

When the Restore Privacy Tunnel client opens ``https://restoreprivacy.online/AUDIT.md``,
it must **not** dump user browsing stats onto the public document. Instead the
client shows two labelled halves:

* **Left — Your browsing stats** (dynamic): dedicated ping + this device's
  connection/visit log. Updates without a full-page reload.
* **Right — Project and files** (manual): public AUDIT.md + project file index.
  Changes only when the user refreshes that pane.

Visit events are appended to the **on-device** connection log only. This module
never uploads log content.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import Any, Iterable, Mapping, Sequence
from urllib.parse import urlparse

# Distinct pane identities (tests + UI must use these exact labels).
LEFT_PANE_ID = "user_browsing_stats"
RIGHT_PANE_ID = "project_files"
LEFT_PANE_LABEL = "Your browsing stats"
RIGHT_PANE_LABEL = "Project and files"

AUDIT_STATUS_PATH = "/AUDIT.md"
DEFAULT_AUDIT_URL = "https://restoreprivacy.online/AUDIT.md"

DEVICE_ONLY_RETENTION_PREFIX = "your data is only retained by your own device. "
DEVICE_ONLY_RETENTION_TYPEWRITER = "privacy, restored."
DEVICE_ONLY_RETENTION_SENTENCE = (
    DEVICE_ONLY_RETENTION_PREFIX + DEVICE_ONLY_RETENTION_TYPEWRITER
)

# Local connection-log kind for an in-client AUDIT visit (device only).
KIND_AUDIT_VISIT = "audit_visit"
AUDIT_VISIT_MESSAGE = "AUDIT.md visit (device-only log)"

WEEKLY_WIPE_MAX_AGE_SEC = 7 * 24 * 3600

_STATE_ORDER = {"Green": 0, "Amber": 1, "Red": 2}


def is_audit_url(url: str) -> bool:
    """True when *url* is the public AUDIT.md document (any origin path)."""
    raw = (url or "").strip()
    if not raw:
        return False
    try:
        parsed = urlparse(raw)
        path = (parsed.path or raw).rstrip("/")
    except ValueError:
        path = raw.rstrip("/")
    return path.lower().endswith("/audit.md") or path.lower() == "audit.md"


def pane_may_auto_update(pane_id: str) -> bool:
    """Left pane is dynamic; right pane is manual-refresh only."""
    return str(pane_id or "").strip() == LEFT_PANE_ID


def pane_label(pane_id: str) -> str:
    pid = str(pane_id or "").strip()
    if pid == LEFT_PANE_ID:
        return LEFT_PANE_LABEL
    if pid == RIGHT_PANE_ID:
        return RIGHT_PANE_LABEL
    return pid or "(unknown pane)"


@dataclass
class AuditSplitState:
    """In-memory two-pane snapshot (not uploaded)."""

    left_sample: dict[str, Any] = field(default_factory=dict)
    right_snapshot: dict[str, Any] = field(default_factory=dict)
    left_generation: int = 0
    right_generation: int = 0

    def to_dict(self) -> dict[str, Any]:
        return {
            "left_sample": dict(self.left_sample),
            "right_snapshot": dict(self.right_snapshot),
            "left_generation": int(self.left_generation),
            "right_generation": int(self.right_generation),
            "left_label": LEFT_PANE_LABEL,
            "right_label": RIGHT_PANE_LABEL,
        }


def apply_pane_refresh(
    state: AuditSplitState | None,
    pane_id: str,
    payload: Mapping[str, Any] | None,
    *,
    explicit: bool = False,
) -> AuditSplitState:
    """Apply a new sample to one pane.

    Left always accepts a new sample. Right accepts a sample **only** when
    *explicit* is True (manual refresh). Opening the view should pass
    ``explicit=True`` once per pane to seed both sides.
    """
    cur = state if isinstance(state, AuditSplitState) else AuditSplitState()
    pid = str(pane_id or "").strip()
    sample = dict(payload or {})
    if pid == RIGHT_PANE_ID and not explicit:
        return cur
    if pid == LEFT_PANE_ID:
        return replace(
            cur,
            left_sample=sample,
            left_generation=int(cur.left_generation) + 1,
        )
    if pid == RIGHT_PANE_ID:
        return replace(
            cur,
            right_snapshot=sample,
            right_generation=int(cur.right_generation) + 1,
        )
    return cur


def catalog_overall_for_installed(
    packages: Sequence[Mapping[str, Any]] | None,
    installed_platforms: Iterable[str] | None,
) -> str:
    """Worst RAG among packages the user actually has installed.

    Uninstalled platforms are ignored so a single-OS device does not paint
    catalog overall Red for other catalog installers they never installed.
    With no installed list, returns Green (do not invent Red).
    """
    installed = {
        str(p or "").strip().lower()
        for p in (installed_platforms or ())
        if str(p or "").strip()
    }
    rows = list(packages or [])
    if not installed:
        return "Green"
    considered = [
        r
        for r in rows
        if str(r.get("platform") or "").strip().lower() in installed
    ]
    if not considered:
        return "Green"
    worst = "Green"
    for row in considered:
        st = str(row.get("state") or "Red")
        if _STATE_ORDER.get(st, 2) > _STATE_ORDER.get(worst, 0):
            worst = st
    return worst


def evaluate_weekly_wipe_prerequisite(
    *,
    last_wipe_at: float | int | None = None,
    now: float | int | None = None,
    max_age_sec: float = WEEKLY_WIPE_MAX_AGE_SEC,
) -> dict[str, Any]:
    """Fail-closed: residual node leak wipe must have run within *max_age_sec*.

    Missing or stale last-wipe is **not** a pass. Weekly erasure of residual
    leak that might be left on the node is an audit prerequisite.
    """
    import time

    t_now = float(now if now is not None else time.time())
    if last_wipe_at is None:
        return {
            "ok": False,
            "prerequisite": "weekly_residual_leak_wipe",
            "reason": "no weekly residual leak wipe recorded",
            "last_wipe_at": None,
            "max_age_sec": float(max_age_sec),
        }
    try:
        last = float(last_wipe_at)
    except (TypeError, ValueError):
        return {
            "ok": False,
            "prerequisite": "weekly_residual_leak_wipe",
            "reason": "weekly residual leak wipe timestamp unreadable",
            "last_wipe_at": last_wipe_at,
            "max_age_sec": float(max_age_sec),
        }
    age = t_now - last
    if age < 0:
        age = 0.0
    if age > float(max_age_sec):
        return {
            "ok": False,
            "prerequisite": "weekly_residual_leak_wipe",
            "reason": (
                f"weekly residual leak wipe is stale ({int(age)}s old; "
                f"max {int(max_age_sec)}s)"
            ),
            "last_wipe_at": last,
            "age_sec": age,
            "max_age_sec": float(max_age_sec),
        }
    return {
        "ok": True,
        "prerequisite": "weekly_residual_leak_wipe",
        "reason": "weekly residual leak wipe current",
        "last_wipe_at": last,
        "age_sec": age,
        "max_age_sec": float(max_age_sec),
    }


def last_wipe_at_from_state(state: Mapping[str, Any] | None) -> float | None:
    """Extract a unix timestamp from fleet-wipe / weekly-wipe state."""
    if not isinstance(state, Mapping):
        return None
    for key in ("last_wipe_at", "last_completed_at", "wiped_at", "completed_at"):
        raw = state.get(key)
        if raw is None or raw == "":
            continue
        if isinstance(raw, (int, float)):
            return float(raw)
        text = str(raw).strip()
        if not text:
            continue
        try:
            return float(text)
        except ValueError:
            pass
        # ISO-8601 Z
        try:
            from datetime import datetime, timezone

            iso = text.replace("Z", "+00:00")
            return datetime.fromisoformat(iso).timestamp()
        except ValueError:
            continue
    return None


def build_user_browsing_stats(
    events: Sequence[Mapping[str, Any] | Any] | None,
    ping: Mapping[str, Any] | None,
    *,
    now: float | None = None,
    platform: str = "",
) -> dict[str, Any]:
    """Left-pane payload from **local** events + dedicated ping (no upload)."""
    import time

    rows: list[dict[str, Any]] = []
    for ev in events or ():
        if isinstance(ev, Mapping):
            rows.append(dict(ev))
        else:
            rows.append(
                {
                    "ts": getattr(ev, "ts", 0),
                    "kind": getattr(ev, "kind", ""),
                    "message": getattr(ev, "message", ""),
                }
            )
    ping = dict(ping or {})
    last_connect = ""
    last_visit = ""
    visit_count = 0
    for ev in rows:
        kind = str(ev.get("kind") or "")
        msg = str(ev.get("message") or "")
        if kind == "connect":
            last_connect = msg
        if kind == KIND_AUDIT_VISIT:
            visit_count += 1
            last_visit = msg
    return {
        "pane": LEFT_PANE_ID,
        "label": LEFT_PANE_LABEL,
        "dynamic": True,
        "platform": str(platform or ""),
        "ping_ok": bool(ping.get("ok")),
        "ping_ms": ping.get("rtt_ms") if ping.get("rtt_ms") is not None else ping.get("rttMs"),
        "ping_host": str(ping.get("host") or ""),
        "ping_error": str(ping.get("error") or ""),
        "event_count": len(rows),
        "audit_visit_count": visit_count,
        "last_connect": last_connect,
        "last_visit": last_visit,
        "recent": [
            {
                "kind": str(e.get("kind") or ""),
                "message": str(e.get("message") or "")[:200],
            }
            for e in rows[-8:]
        ],
        "retention": DEVICE_ONLY_RETENTION_SENTENCE,
        "retention_prefix": DEVICE_ONLY_RETENTION_PREFIX,
        "retention_typewriter": DEVICE_ONLY_RETENTION_TYPEWRITER,
        "sampled_at": float(now if now is not None else time.time()),
        "device_only": True,
    }


def build_project_files_snapshot(
    *,
    audit_text: str = "",
    file_names: Sequence[str] | None = None,
    catalog_overall: str = "",
    catalog_version: str = "",
) -> dict[str, Any]:
    """Right-pane payload: project AUDIT + file names (no user browsing stats)."""
    names = [str(n).strip() for n in (file_names or ()) if str(n).strip()]
    body = str(audit_text or "")
    return {
        "pane": RIGHT_PANE_ID,
        "label": RIGHT_PANE_LABEL,
        "dynamic": False,
        "manual_refresh_only": True,
        "catalog_version": str(catalog_version or ""),
        "catalog_overall": str(catalog_overall or ""),
        "audit_excerpt": body[:4000],
        "audit_chars": len(body),
        "files": names,
        "retention": DEVICE_ONLY_RETENTION_SENTENCE,
    }


def default_project_file_names() -> list[str]:
    """Main project documents shown on the right pane."""
    return [
        "AUDIT.md",
        "README.md",
        "PRIVACY_POLICY.md",
        "LICENSE",
        "CREDITS.md",
        "client/VERSION",
    ]


def visit_log_detail(*, platform: str = "", ping_ms: Any = None) -> dict[str, Any]:
    """Safe local-only detail for an AUDIT visit append."""
    out: dict[str, Any] = {
        "surface": "AUDIT.md",
        "device_only": True,
        "uploaded": False,
    }
    if platform:
        out["platform"] = str(platform)
    if ping_ms is not None:
        try:
            out["ping_ms"] = float(ping_ms)
        except (TypeError, ValueError):
            pass
    return out


def append_audit_visit_to_device_log(
    *,
    path: Any = None,
    platform: str = "",
    ping_ms: Any = None,
    ts: float | None = None,
) -> Any:
    """Append this AUDIT visit to the **on-device** connection log (never upload).

    Imperative: a user opening AUDIT.md in the RPT client must leave a local
    visit record on their device.
    """
    from client.connection_log import append_event

    return append_event(
        KIND_AUDIT_VISIT,
        AUDIT_VISIT_MESSAGE,
        path=path,
        ts=ts,
        detail=visit_log_detail(platform=platform, ping_ms=ping_ms),
    )


def render_split_markup(state: AuditSplitState) -> str:
    """Structural HTML used by the in-client view (and source-excerpt tests)."""
    left = state.left_sample or {}
    right = state.right_snapshot or {}
    ping = left.get("ping_ms")
    ping_txt = f"{ping} ms" if ping is not None else "n/a"
    files = right.get("files") or []
    file_lis = "".join(f"<li>{_esc(str(n))}</li>" for n in files)
    recent = left.get("recent") or []
    rec_lis = "".join(
        f"<li>{_esc(str(r.get('kind','')))}: {_esc(str(r.get('message','')))}</li>"
        for r in recent
        if isinstance(r, Mapping)
    )
    prefix = _esc(DEVICE_ONLY_RETENTION_PREFIX)
    typed = _esc(DEVICE_ONLY_RETENTION_TYPEWRITER)
    return f"""<section id="rpt-audit-split" data-audit-split="1">
  <aside id="rpt-audit-left" data-pane="{LEFT_PANE_ID}" data-dynamic="1"
         aria-label="{LEFT_PANE_LABEL}">
    <h2 class="rpt-audit-pane-label">{LEFT_PANE_LABEL}</h2>
    <p class="rpt-audit-ping">dedicated ping: {_esc(ping_txt)}</p>
    <ul class="rpt-audit-recent">{rec_lis}</ul>
    <p class="rpt-audit-retention">{prefix}<span class="suite-typewriter suite-typewriter-welcome neon-type"
       data-typewriter="1" data-typewriter-role="welcome"
       data-typewriter-text="{typed}">{typed}</span></p>
  </aside>
  <main id="rpt-audit-right" data-pane="{RIGHT_PANE_ID}" data-dynamic="0"
        data-manual-refresh="1" aria-label="{RIGHT_PANE_LABEL}">
    <h2 class="rpt-audit-pane-label">{RIGHT_PANE_LABEL}</h2>
    <p class="rpt-audit-catalog">catalog { _esc(str(right.get('catalog_version') or '')) }
       overall {_esc(str(right.get('catalog_overall') or ''))}</p>
    <ul class="rpt-audit-files">{file_lis}</ul>
    <button type="button" id="rpt-audit-right-refresh">Refresh project files</button>
    <pre class="rpt-audit-md">{_esc(str(right.get('audit_excerpt') or '')[:1200])}</pre>
  </main>
</section>
"""


def show_audit_split_window(
    parent: Any,
    *,
    connection_log_path: Any = None,
    platform: str = "",
    audit_text: str = "",
    ping: Mapping[str, Any] | None = None,
) -> AuditSplitState:
    """Open the in-client two-pane AUDIT view and **append the visit to device log**.

    Tk window when *parent* is a Tk widget; otherwise still appends the visit
    and returns the seeded pane state (headless / tests).
    """
    ev = append_audit_visit_to_device_log(
        path=connection_log_path,
        platform=platform,
        ping_ms=(ping or {}).get("rtt_ms"),
    )
    try:
        from client.connection_log import read_events

        events = read_events(path=connection_log_path)
    except Exception:
        events = [ev]
    left = build_user_browsing_stats(events, ping, platform=platform)
    right = build_project_files_snapshot(
        audit_text=audit_text,
        file_names=default_project_file_names(),
        catalog_overall=catalog_overall_for_installed(
            [{"platform": platform, "state": "Green"}],
            [platform] if platform else [],
        ),
    )
    state = apply_pane_refresh(None, LEFT_PANE_ID, left, explicit=True)
    state = apply_pane_refresh(state, RIGHT_PANE_ID, right, explicit=True)
    try:
        import tkinter as tk
        from tkinter import ttk
    except Exception:
        return state
    if parent is None:
        return state
    try:
        win = tk.Toplevel(parent)
        win.title("Most recent audit")
        win.geometry("960x560")
        pane = ttk.Panedwindow(win, orient=tk.HORIZONTAL)
        pane.pack(fill=tk.BOTH, expand=True)
        left_fr = ttk.Frame(pane, padding=12)
        right_fr = ttk.Frame(pane, padding=12)
        pane.add(left_fr, weight=1)
        pane.add(right_fr, weight=1)
        ttk.Label(left_fr, text=LEFT_PANE_LABEL, font=("Segoe UI", 12, "bold")).pack(
            anchor="w"
        )
        ttk.Label(
            left_fr,
            text=f"dedicated ping: {left.get('ping_ms') if left.get('ping_ms') is not None else 'n/a'}",
        ).pack(anchor="w", pady=(8, 0))
        ttk.Label(
            left_fr,
            text=f"visits on this device: {left.get('audit_visit_count') or 0}",
        ).pack(anchor="w")
        ttk.Label(
            left_fr,
            text=DEVICE_ONLY_RETENTION_PREFIX,
            wraplength=400,
            justify=tk.LEFT,
        ).pack(anchor="w", pady=(16, 0))
        ttk.Label(
            left_fr,
            text=DEVICE_ONLY_RETENTION_TYPEWRITER,
            font=("Segoe UI", 11, "bold"),
        ).pack(anchor="w")
        ttk.Label(right_fr, text=RIGHT_PANE_LABEL, font=("Segoe UI", 12, "bold")).pack(
            anchor="w"
        )
        for name in default_project_file_names():
            ttk.Label(right_fr, text=name).pack(anchor="w")

        def _manual_right() -> None:
            nonlocal state
            nxt = build_project_files_snapshot(
                audit_text=audit_text,
                file_names=default_project_file_names(),
            )
            state = apply_pane_refresh(state, RIGHT_PANE_ID, nxt, explicit=True)

        ttk.Button(
            right_fr, text="Refresh project files", command=_manual_right
        ).pack(anchor="w", pady=(12, 0))
    except Exception:
        return state
    return state


def _esc(text: str) -> str:
    return (
        str(text)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
    )

File: client_app/lib/audit_split_view.dart
/// In-client AUDIT.md split view: left browsing stats (dynamic), right project files (manual).
///
/// Opening this surface **always** appends a visit event to the on-device
/// connection log. Nothing is uploaded.
library;

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';

import 'connection_log.dart';
import 'legal_links.dart';
import 'node_ping.dart';
import 'rpt_config.dart';
import 'suite_version.dart';
import 'theme.dart';

const String kLeftPaneId = 'user_browsing_stats';
const String kRightPaneId = 'project_files';
const String kLeftPaneLabel = 'Your browsing stats';
const String kRightPaneLabel = 'Project and files';

const String kDeviceOnlyRetentionPrefix =
    'your data is only retained by your own device. ';
const String kDeviceOnlyRetentionTypewriter = 'privacy, restored.';
const String kDeviceOnlyRetentionSentence =
    '$kDeviceOnlyRetentionPrefix$kDeviceOnlyRetentionTypewriter';

const String kAuditVisitKind = 'audit_visit';
const String kAuditVisitMessage = 'AUDIT.md visit (device-only log)';

const List<String> kDefaultProjectFileNames = [
  'AUDIT.md',
  'README.md',
  'PRIVACY_POLICY.md',
  'LICENSE',
  'CREDITS.md',
  'client/VERSION',
];

const Key kAuditSplitViewKey = Key('audit_split_view');
const Key kAuditLeftPaneKey = Key('audit_left_pane');
const Key kAuditRightPaneKey = Key('audit_right_pane');
const Key kAuditRightRefreshKey = Key('audit_right_refresh');

bool paneMayAutoUpdate(String paneId) => paneId == kLeftPaneId;

/// Apply a sample to one pane. Right side ignores implicit (non-explicit) updates.
Map<String, dynamic> applyPaneRefresh(
  Map<String, dynamic> state,
  String paneId,
  Map<String, dynamic> payload, {
  bool explicit = false,
}) {
  final next = Map<String, dynamic>.from(state);
  if (paneId == kRightPaneId && !explicit) {
    return next;
  }
  if (paneId == kLeftPaneId) {
    next['left_sample'] = Map<String, dynamic>.from(payload);
    next['left_generation'] = (next['left_generation'] as int? ?? 0) + 1;
  } else if (paneId == kRightPaneId) {
    next['right_snapshot'] = Map<String, dynamic>.from(payload);
    next['right_generation'] = (next['right_generation'] as int? ?? 0) + 1;
  }
  return next;
}

/// Worst RAG among packages this device actually has installed.
String catalogOverallForInstalled(
  List<Map<String, String>> packages,
  List<String> installedPlatforms,
) {
  const order = {'Green': 0, 'Amber': 1, 'Red': 2};
  final installed = installedPlatforms
      .map((p) => p.trim().toLowerCase())
      .where((p) => p.isNotEmpty)
      .toSet();
  if (installed.isEmpty) return 'Green';
  final considered = packages.where((p) {
    return installed.contains((p['platform'] ?? '').toLowerCase());
  });
  var worst = 'Green';
  for (final row in considered) {
    final st = row['state'] ?? 'Red';
    if ((order[st] ?? 2) > (order[worst] ?? 0)) worst = st;
  }
  return worst;
}

Map<String, dynamic> buildUserBrowsingStats({
  required List<ConnectionLogEvent> events,
  PingResult? ping,
  String platform = '',
  double? now,
}) {
  var visitCount = 0;
  var lastVisit = '';
  var lastConnect = '';
  for (final e in events) {
    if (e.kind == kAuditVisitKind) {
      visitCount += 1;
      lastVisit = e.message;
    }
    if (e.kind == kLogKindConnect) lastConnect = e.message;
  }
  return {
    'pane': kLeftPaneId,
    'label': kLeftPaneLabel,
    'dynamic': true,
    'platform': platform,
    'ping_ok': ping?.ok ?? false,
    'ping_ms': ping?.rttMs,
    'ping_host': ping?.host ?? '',
    'event_count': events.length,
    'audit_visit_count': visitCount,
    'last_connect': lastConnect,
    'last_visit': lastVisit,
    'retention': kDeviceOnlyRetentionSentence,
    'device_only': true,
    'sampled_at': now ?? DateTime.now().millisecondsSinceEpoch / 1000.0,
  };
}

Map<String, dynamic> buildProjectFilesSnapshot({
  String auditText = '',
  List<String> fileNames = const [],
  String catalogOverall = '',
  String catalogVersion = '',
}) {
  return {
    'pane': kRightPaneId,
    'label': kRightPaneLabel,
    'dynamic': false,
    'manual_refresh_only': true,
    'catalog_version': catalogVersion,
    'catalog_overall': catalogOverall,
    'audit_excerpt': auditText.length > 4000 ? auditText.substring(0, 4000) : auditText,
    'files': List<String>.from(fileNames),
  };
}

/// Fetch the public project AUDIT.md (no user stats). Used on first open and
/// every explicit right-pane refresh so the project half can change.
Future<String> fetchPublicAuditMarkdown({
  Uri? uri,
  Future<String> Function(Uri url)? getText,
}) async {
  final target = uri ?? Uri.parse('${LegalDocLink.statusOrigin}/AUDIT.md');
  if (getText != null) {
    return getText(target);
  }
  final client = HttpClient();
  try {
    final req = await client.getUrl(target);
    req.headers.set(HttpHeaders.userAgentHeader, 'rpt-client-audit-split/1');
    final resp = await req.close();
    if (resp.statusCode < 200 || resp.statusCode >= 300) {
      return '';
    }
    return await resp.transform(utf8.decoder).join();
  } catch (_) {
    return '';
  } finally {
    client.close(force: true);
  }
}

/// Load the right-pane project snapshot (AUDIT.md + file names).
Future<Map<String, dynamic>> loadProjectAuditSnapshot({
  String platform = '',
  String catalogVersion = kSuiteVersion,
  Future<String> Function()? fetchAuditText,
  List<String>? fileNames,
}) async {
  final text = fetchAuditText != null
      ? await fetchAuditText()
      : await fetchPublicAuditMarkdown();
  return buildProjectFilesSnapshot(
    auditText: text,
    fileNames: fileNames ?? kDefaultProjectFileNames,
    catalogVersion: catalogVersion,
    catalogOverall: catalogOverallForInstalled(
      [
        {'platform': platform, 'state': 'Green'},
      ],
      platform.isEmpty ? const <String>[] : [platform],
    ),
  );
}

/// Imperative: record this AUDIT visit on the **device** connection log.
Future<ConnectionLogEvent> appendAuditVisitToDeviceLog(
  ConnectionLog log, {
  String platform = '',
  double? pingMs,
}) {
  final detail = <String, String>{
    'surface': 'AUDIT.md',
    'device_only': 'true',
    'uploaded': 'false',
  };
  if (platform.isNotEmpty) detail['platform'] = platform;
  if (pingMs != null) detail['ping_ms'] = pingMs.toString();
  return log.appendEvent(kAuditVisitKind, kAuditVisitMessage, detail: detail);
}

/// In-client two-pane AUDIT surface (opened instead of the public AUDIT.md URL).
class AuditSplitView extends StatefulWidget {
  const AuditSplitView({
    super.key,
    required this.connectionLog,
    this.platformLabel = 'flutter',
    this.multihopOn = false,
    this.auditText = '',
    this.fileNames = kDefaultProjectFileNames,
    this.leftPoll,
    this.pingProbe,
    this.projectSnapshotLoader,
    this.fetchAuditText,
  });

  final ConnectionLog connectionLog;
  final String platformLabel;
  final bool multihopOn;
  final String auditText;
  final List<String> fileNames;

  /// Injected ticker for tests (left-pane dynamic refresh).
  final Stream<void>? leftPoll;
  final Future<PingResult> Function(String host)? pingProbe;

  /// Right pane: called on first open and every explicit refresh.
  final Future<Map<String, dynamic>> Function()? projectSnapshotLoader;
  final Future<String> Function()? fetchAuditText;

  @override
  State<AuditSplitView> createState() => _AuditSplitViewState();
}

class _AuditSplitViewState extends State<AuditSplitView> {
  Map<String, dynamic> _state = {
    'left_sample': <String, dynamic>{},
    'right_snapshot': <String, dynamic>{},
    'left_generation': 0,
    'right_generation': 0,
  };
  StreamSubscription<void>? _poll;
  Timer? _timer;
  bool _typed = false;

  @override
  void initState() {
    super.initState();
    _openVisit();
    final poll = widget.leftPoll;
    if (poll != null) {
      _poll = poll.listen((_) => _refreshLeft());
    } else {
      _timer = Timer.periodic(const Duration(seconds: 8), (_) => _refreshLeft());
    }
    Future<void>.delayed(const Duration(milliseconds: 80), () {
      if (mounted) setState(() => _typed = true);
    });
  }

  @override
  void dispose() {
    _poll?.cancel();
    _timer?.cancel();
    super.dispose();
  }

  Future<void> _openVisit() async {
    final ping = await _probe();
    await appendAuditVisitToDeviceLog(
      widget.connectionLog,
      platform: widget.platformLabel,
      pingMs: ping.rttMs,
    );
    await _refreshLeft(ping: ping);
    await _refreshRight();
  }

  Future<PingResult> _probe() async {
    final fn = widget.pingProbe;
    if (fn != null) return fn(RptConfig.entryHost);
    try {
      final r = await measureSettingsPings(
        multihopOn: widget.multihopOn,
        probe: widget.pingProbe,
      );
      return r.entry;
    } catch (_) {
      return const PingResult(host: '', port: kStatusTcpPort, ok: false, error: 'probe_failed');
    }
  }

  Future<void> _refreshLeft({PingResult? ping}) async {
    final p = ping ?? await _probe();
    final events = await widget.connectionLog.readEvents(limit: 40);
    final sample = buildUserBrowsingStats(
      events: events,
      ping: p,
      platform: widget.platformLabel,
    );
    if (!mounted) return;
    setState(() {
      _state = applyPaneRefresh(_state, kLeftPaneId, sample);
    });
  }

  Future<void> _refreshRight() async {
    final loader = widget.projectSnapshotLoader;
    final Map<String, dynamic> snap;
    if (loader != null) {
      snap = await loader();
    } else {
      snap = await loadProjectAuditSnapshot(
        platform: widget.platformLabel,
        catalogVersion: kSuiteVersion,
        fetchAuditText: widget.fetchAuditText ??
            (widget.auditText.isEmpty ? null : () async => widget.auditText),
        fileNames: widget.fileNames,
      );
    }
    if (!mounted) return;
    setState(() {
      _state = applyPaneRefresh(_state, kRightPaneId, snap, explicit: true);
    });
  }

  @override
  Widget build(BuildContext context) {
    final left = Map<String, dynamic>.from(_state['left_sample'] as Map? ?? {});
    final right = Map<String, dynamic>.from(_state['right_snapshot'] as Map? ?? {});
    final pingMs = left['ping_ms'];
    final files = (right['files'] as List?)?.map((e) => e.toString()).toList() ?? const <String>[];
    return Scaffold(
      key: kAuditSplitViewKey,
      backgroundColor: suiteChromeBgOf(context),
      appBar: AppBar(
        title: const Text('Most recent audit'),
        backgroundColor: suitePanelBgOf(context),
      ),
      body: Row(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Expanded(
            child: Container(
              key: kAuditLeftPaneKey,
              padding: const EdgeInsets.all(16),
              color: suitePanelBgOf(context),
              child: ListView(
                children: [
                  Text(
                    kLeftPaneLabel,
                    style: TextStyle(
                      color: suitePrimaryOf(context),
                      fontWeight: FontWeight.w700,
                      fontSize: 16,
                    ),
                  ),
                  const SizedBox(height: 8),
                  Text(
                    'dedicated ping: ${pingMs == null ? 'n/a' : '$pingMs ms'}',
                    style: TextStyle(color: suiteTextOf(context)),
                  ),
                  const SizedBox(height: 8),
                  Text(
                    'visits on this device: ${left['audit_visit_count'] ?? 0}',
                    style: TextStyle(color: suiteTextOf(context)),
                  ),
                  const SizedBox(height: 16),
                  Text.rich(
                    TextSpan(
                      style: TextStyle(color: suiteTextOf(context), fontSize: 14),
                      children: [
                        const TextSpan(text: kDeviceOnlyRetentionPrefix),
                        TextSpan(
                          text: _typed ? kDeviceOnlyRetentionTypewriter : '',
                          style: TextStyle(
                            color: suitePrimaryOf(context),
                            fontWeight: FontWeight.w600,
                          ),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
            ),
          ),
          const VerticalDivider(width: 1),
          Expanded(
            child: Container(
              key: kAuditRightPaneKey,
              padding: const EdgeInsets.all(16),
              child: ListView(
                children: [
                  Text(
                    kRightPaneLabel,
                    style: TextStyle(
                      color: suitePrimaryOf(context),
                      fontWeight: FontWeight.w700,
                      fontSize: 16,
                    ),
                  ),
                  const SizedBox(height: 8),
                  Text(
                    'catalog ${right['catalog_version'] ?? ''} overall ${right['catalog_overall'] ?? ''}',
                    style: TextStyle(color: suiteTextMutedOf(context)),
                  ),
                  const SizedBox(height: 8),
                  ...files.map(
                    (n) => Text(n, style: TextStyle(color: suiteTextOf(context))),
                  ),
                  const SizedBox(height: 12),
                  Align(
                    alignment: Alignment.centerLeft,
                    child: TextButton(
                      key: kAuditRightRefreshKey,
                      onPressed: _refreshRight,
                      child: const Text('Refresh project files'),
                    ),
                  ),
                  const SizedBox(height: 8),
                  Text(
                    (right['audit_excerpt'] ?? '').toString(),
                    style: TextStyle(
                      color: suiteTextMutedOf(context),
                      fontFamily: 'monospace',
                      fontSize: 11,
                    ),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}
